step/src/main/java/com/google/sps/MatchManager.java

```diff
@@ -17,50 +17,113 @@
+  public static void addUserToMatchQueue(User u) {
+      matchQueue.add(u);
+  }
+
+  // Matchmaker -- takes in the user who just requested a match and matches them
+  // with the one who is in the queue having the highest number of mutual interests with the person
+  // or adds them to the
   // queue to wait to be matched if no on else is in there
   public static void generateMatch(User user) {
     user.updateMatchPendingStatus(true);
 
     if (matchQueue.isEmpty()) {
-      matchQueue.add(user);  
-    } else {
-      User matchResult = matchQueue.peek();
+        matchQueue.add(user); 
+        return;
+    }
+    matchQueue.add(user); 
+
+    if (!matchQueue.isEmpty()) {
+      matchQueue.remove(user);
+      User matchResult = findCompatibleMatch(user);
+      if (matchResult == null) {
+          matchQueue.add(user); 
+          user.updateMatchPendingStatus(true);
+          return;
+      }
 
       if (!matchResult.isMatchedWith(user) && !matchResult.equals(user)) {
         // actually removing the match result from the queue
-        matchResult = matchQueue.poll(); 
         createMatch(matchResult, user);
+        matchQueue.remove(matchResult);  
+        
       // adding the user to the queue in case there wasn't a successful match for them
       // this is so they can be matched later      
-      } else if (!matchQueue.contains(user)) {
-        matchQueue.add(user);  
+      } else if (!matchQueue.contains(user)) {  
+          matchQueue.add(user); 
       }  
     }    
   }
 
+  // Match the user with the most compatible user in the database (the one who has the most number
+  // of similar interests). 
+  // Currently, the minimum percentage of mutual interests between 2 users required to get matched are 60%
+  public static User findCompatibleMatch(User firstUser) {
+    HashMap<User, Integer> mutualInterests = new HashMap<>();
+    for (User secondUser: matchQueue) {
+        mutualInterests.put(secondUser, 0);
+    } 
+    for (User secondUser: matchQueue) {
+        for (int i = 0; i < secondUser.getPreferences().size(); i++) {
+            // Compare the second user's each preference answer with the first user's
+            if (firstUser.getPreferences().get(i).equals(secondUser.getPreferences().get(i))) {
+                mutualInterests.replace(secondUser, mutualInterests.get(secondUser)+1);
+            }
+        }
+    }
+
+    //Traverse through the hashmap to find the user with the maximum number of mutual interests with the first user
+    Iterator iterator = mutualInterests.entrySet().iterator();
+    int max = 0; 
+    String compatibleUserId = "";
+
+    while (iterator.hasNext()) {
+        Map.Entry mapE = (Map.Entry)iterator.next();
+        // FIFO
+        if (max < (int)mapE.getValue()) {
+            max = (int) mapE.getValue();
+            compatibleUserId = ((User)mapE.getKey()).getId();
+        }
+    }
+
+    if (max < MINIMUM_INTERESTS_PERCENTAGE_REQUIRED * QUESTIONS) {
```

Comment:
This might get tricky given that users currently have the option of leaving questions blank. We might want to go back and make the preference questions mandatory

Reply:
I don't think we should make them mandatory. Users shouldn't have to provide data they don't want to and if we make the questions mandatory a user who doesn't want to answer a question will probably just put fake answers which will create a lower quality match.

Reply:
I will figure out if there is a way to make the questions required and put another answer like "don't want to answer"

Reply:
Yes, but it might be hard to find users who have 60% match if users are leaving answers blank. Matches might also not be that good cause people who leave a lot of answers blank will just end up getting matched to one another.